### PR TITLE
Auth mechanism

### PR DESCRIPTION
**Pull request by skosinskiy**

### CODE REVIEW

**Comment by skosinskiy** [comment]
[5SP - Implement auth mechanism](https://trello.com/c/BehvIqJ6/69-5sp-implement-auth-mechanism)

File: frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js [hunk 1]
```diff
@@ -1,30 +1,43 @@
...
-          <Route path="/forgot-password" component={ForgotPassword}/>
-          <Route path="/reset-password" component={ResetPassword} />
-          <Route path="/" component={Main}/>
-        </Switch>
-      )
-    }
+  render () {
+    const {isAuthenticated} = this.props
+
+    // if (isAuthenticated) {
```
Comment (zozich) [inline frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js]: Remove commented code

File: frontend/admin_panel/src/reducers/users/index.js [hunk 2]
```diff
@@ -4,9 +4,9 @@ const initialState = {
   usersListByEmail: [],
   userRoles: [],
   changedUsersList: new Set(),
-  isAuthenticated: true,
+  isAuthenticated: false,
```
Comment (zozich) [inline frontend/admin_panel/src/reducers/users/index.js]: We don't need this variable, please remove it from Redox store. It should be calculated dynamically

File: frontend/admin_panel/src/actions/users/index.js [hunk 3]
```diff
@@ -17,12 +17,31 @@ export const saveUserRoles = (userId, roles) => dispatch => {
...
+    .catch(() => dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {isCurrentUserLoading: false}}))
+}
+
 export const getCurrentUser = () => dispatch => {
-  dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {loading: true}})
+  dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {isCurrentUserLoading: true}})
+  fetchCurrentUser(dispatch)
+}
+
+const fetchCurrentUser = (dispatch) => {
```
Comment (zozich) [inline frontend/admin_panel/src/actions/users/index.js]: This function is not needed

File: frontend/admin_panel/src/actions/users/index.js [hunk 4]
```diff
@@ -17,12 +17,31 @@ export const saveUserRoles = (userId, roles) => dispatch => {
...
   api.get('/api/users/current')
     .then(user => {
-      dispatch({type: Actions.Users.CURRENT_USER_FETCHED, payload: {currentUser: user}})
-      dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {loading: false}})
+      if (user !== '') {
+        dispatch({type: Actions.Users.CURRENT_USER_FETCHED, payload: {currentUser: user}})
+        dispatch({type: Actions.Users.AUTHENTICATE_USER, payload: {isAuthenticated: true}})
```
Comment (zozich) [inline frontend/admin_panel/src/actions/users/index.js]: We need only one action - to store current user in Redux

File: frontend/admin_panel/src/actions/users/index.js [hunk 5]
```diff
@@ -17,12 +17,31 @@ export const saveUserRoles = (userId, roles) => dispatch => {
...
+    if (res.status === 200) {
+      fetchCurrentUser(dispatch)
+    }
+  })
+    .catch(() => dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {isCurrentUserLoading: false}}))
+}
+
 export const getCurrentUser = () => dispatch => {
-  dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {loading: true}})
+  dispatch({type: Actions.Users.CURRENT_USER_LOADING, payload: {isCurrentUserLoading: true}})
```
Comment (zozich) [inline frontend/admin_panel/src/actions/users/index.js]: You can simplify this to `payload: true`. Action does not need to know about name of property in Redux store

File: frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js [hunk 6]
```diff
@@ -1,30 +1,43 @@
...
+    //   return (<RootRoute path="/" authenticated={isAuthenticated}/>)
+    // }
+    
+    return (
+      <Switch>
+        <Route path="/login" component={Login}/>
+        <ProtectedRoute path="/admin" component={Admin} authenticated={isAuthenticated}/>
```
Comment (zozich) [inline frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js]: Let's make this component mapped on `/`. We don't need `/admin` in routes - it is all Admin panel

File: frontend/admin_panel/src/components/SignIn/index.js [hunk 7]
```diff
@@ -43,60 +46,93 @@ const styles = theme => ({
...
   }
 })
 
-function SignIn (props) {
-  const { classes } = props
+class SignIn extends Component {
+  render () {
+    const { classes, isAuthenticated, isCurrentUserLoading } = this.props
+    
+    if (isCurrentUserLoading) {
```
Comment (zozich) [inline frontend/admin_panel/src/components/SignIn/index.js]: Do we need this code at all? Loading is managed on App.js level

File: src/main/java/com/danit/finalproject/application/entity/Role.java [hunk 8]
```diff
@@ -27,9 +28,8 @@
   @ManyToMany(mappedBy = "roles")
   private List<User> users;
 
-  @ManyToMany
-  @JoinTable(name = "permissions_roles",
-          joinColumns = {@JoinColumn(name = "role_id")},
-          inverseJoinColumns = {@JoinColumn(name = "permission_id")})
+  @ElementCollection
+  @CollectionTable(name = "role_permissions", joinColumns = @JoinColumn(name = "role_id"))
+  @Column(name = "permission_id")
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/entity/Role.java]: Shouldn't it be `@JoinColumn`?

File: src/main/resources/application-dev.yml [hunk 9]
```diff
@@ -0,0 +1,33 @@
+spring:
```
Comment (zozich) [inline src/main/resources/application-dev.yml]: 1. Let's keep all `yml` configs in a single file
2. All common properties should be stored at a root level and not be duplicated

File: frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js (hunk 1 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js]: done

File: frontend/admin_panel/src/reducers/users/index.js (hunk 2 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/reducers/users/index.js]: done

File: frontend/admin_panel/src/actions/users/index.js (hunk 4 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/actions/users/index.js]: done

File: frontend/admin_panel/src/actions/users/index.js (hunk 3 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/actions/users/index.js]: done

File: frontend/admin_panel/src/actions/users/index.js (hunk 5 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/actions/users/index.js]: done

File: src/main/java/com/danit/finalproject/application/entity/Role.java (hunk 8 above)
Comment (skosinskiy) [reply src/main/java/com/danit/finalproject/application/entity/Role.java]: I've googled such example, the db structure is correct

File: frontend/admin_panel/src/components/SignIn/index.js (hunk 7 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/components/SignIn/index.js]: done

File: frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js (hunk 6 above)
Comment (skosinskiy) [reply frontend/admin_panel/src/components/AppRoutes/MainRouter/index.js]: done

File: src/main/resources/application-dev.yml (hunk 9 above)
Comment (skosinskiy) [reply src/main/resources/application-dev.yml]: done

**Pull request merged**